File: Examples/C/Examples/TcpConnection.c
#include "../../../x-IMU3-API/C/Ximu3.h"
#include "../Helpers.h"
#include "Connection.h"
#include <stdio.h>

void TcpConnection()
{
    if (AskQuestion("Search for connections?"))
    {
        printf("Searching for connections\n");

        XIMU3_NetworkAnnouncement* const networkAnnouncement = XIMU3_network_announcement_new();

        if (XIMU3_network_announcement_get_result(networkAnnouncement) != XIMU3_ResultOk)
        {
            printf("Unable to open network announcement socket\n");
            XIMU3_network_announcement_free(networkAnnouncement);
            return;
        }

        const XIMU3_NetworkAnnouncementMessages messages = XIMU3_network_announcement_get_messages_after_short_delay(networkAnnouncement);

        if (messages.length == 0)
        {
            printf("No TCP connections available\n");
            return;
        }
        printf("Found %s %s\n", messages.array[0].device_name, messages.array[0].serial_number);

        const XIMU3_TcpConnectionInfo connectionInfo = XIMU3_network_announcement_message_to_tcp_connection_info(messages.array[0]);

        XIMU3_network_announcement_messages_free(messages);

        Run(XIMU3_connection_new_tcp(connectionInfo), XIMU3_tcp_connection_info_to_string(connectionInfo));
    }
    else
    {
        const XIMU3_TcpConnectionInfo connectionInfo = (XIMU3_TcpConnectionInfo) {
                .ip_address = "192.168.1.1",
                .port = 7000,
        };
        Run(XIMU3_connection_new_tcp(connectionInfo), XIMU3_tcp_connection_info_to_string(connectionInfo));
    }
}

File: Examples/Rust/src/examples/tcp_connection.rs
use super::connection;
use crate::helpers;
use std::net::Ipv4Addr;
use ximu3::connection_info::*;
use ximu3::network_announcement::*;

pub fn run() {
    if helpers::yes_or_no("Search for connections?") {
        let network_announcement = match NetworkAnnouncement::new() {
            Ok(network_announcement) => network_announcement,
            Err(error) => {
                println!("Network announcement failed. {error}");
                return;
            }
        };

        let messages = network_announcement.get_messages_after_short_delay();

        let Some(message) = messages.first() else {
            println!("No TCP connections available");
            return;
        };

        println!("Found {} {}", message.device_name, message.serial_number);

        let connection_info = &ConnectionInfo::TcpConnectionInfo(message.into());

        connection::run(connection_info);
    } else {
        let connection_info = &ConnectionInfo::TcpConnectionInfo(TcpConnectionInfo {
            ip_address: Ipv4Addr::new(192, 168, 1, 1),
            port: 7000,
        }); // replace with actual connection info

        connection::run(connection_info);
    }
}
